xclim/core/units.py

```diff
@@ -277,56 +283,91 @@ def convert_units_to(
 
     Returns
     -------
-    xr.DataArray or float or int or str or Any
+    str or xr.DataArray or units.Unit or units.Quantity
         The source value converted to target's units.
+        The outputted type is always similar to `source` initial type.
     """
     # Target units
     if isinstance(target, units.Unit):
-        tu = target
+        target_unit = target
     elif isinstance(target, (str, xr.DataArray)):
-        tu = units2pint(target)
+        target_unit = units2pint(target)
     else:
-        raise NotImplementedError
+        raise NotImplementedError(
+            "target must be either a pint Unit or a xarray DataArray."
+        )
 
     if isinstance(source, str):
         q = str2pint(source)
         # Return magnitude of converted quantity. This is going to fail if units are not compatible.
-        return q.to(tu).m
+        return q.to(target_unit).m
 
     if isinstance(source, units.Quantity):
-        return source.to(tu).m
+        return source.to(target_unit).m
 
     if isinstance(source, xr.DataArray):
-        fu = units2pint(source)
-        tu_u = pint2cfunits(tu)
-
-        if fu == tu:
+        source_unit = units2pint(source)
```

Comment:
I have the feeling some of the logic here would fit better within `_rate_and_amount_converter`.